### PR TITLE
Service topology alpha documentation

### PR DESCRIPTION
**Pull request by johnbelamaric**
Placeholder docs PR for https://github.com/kubernetes/enhancements/issues/536

/sig network

### CODE REVIEW

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+## Configuring service topology
+
+Placeholder text.
+
+{{% /capture %}}
+
+{{% capture whatsnext %}}
+
+* [Enabling Service Topology](/docs/tasks/administer-cluster/enabling-service-topology)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/service-topology.md]:
```suggestion
* Read about [enabling Service Topology](/docs/tasks/administer-cluster/enabling-service-topology/)
```

**Comment by imroc** [comment]
Can you add me as collaborator to your repo? I can help to write the docs, push commit to your remote branch.

**Comment by imroc** [comment]
## Prerequisites

The following prerequisites are needed in order to enable topology aware service routing:

* Kubernetes 1.17 or later
* Kube-proxy running in mode IPVS or Iptables
* Enable [Endpoint Slices](/docs/concepts/services-networking/endpoint-slices/) (enabled by default from v1.17)

**Comment by imroc** [comment]
## Enable Service Topology

To enable service topology, enable the `ServiceTopology` feature gate for the kube-proxy:

```
--feature-gates="ServiceTopology=true"
```

**Comment by johnbelamaric** [comment]
You can submit PRs against my fork if you want. Or I can incorporate your suggestions.

**Comment by imroc** [comment]
> You can submit PRs against my fork if you want. Or I can incorporate your suggestions.

I'm going to submit PRs against to your fork.

**Comment by daminisatya** [comment]
@imroc @johnbelamaric Just a reminder about the last Docs deadline - 22nd Nov, by which this PR needs to be merged!

**Comment by imroc** [comment]
Sorry for the delay. I think I can just comment the suggestions to the PR, and @johnbelamaric write to the commits.

**Comment by imroc** [comment]
## Motivation
Figure out a generic way to implement the "local service" route, say "topology aware routing of service",  simply call it "service topology".

Locality is defined by user, it can be any topology-related thing. "Local" means the "same topology level", e.g. same node, same rack, same failure zone, same failure region, same cloud provider etc. Two nodes are considered "local" if they have the same value for a particular label, called the "topology key".

**Comment by imroc** [comment]
## Topology Keys

If your cluster has Service Topology enabled, you can create [Services](https://kubernetes.io/docs/concepts/services-networking/service/) with topology keys by setting a field: `.spec.topologyKeys`.

`topologyKeys` is a preference-order list of topology keys which implementations of services should use to preferentially sort endpoints when accessing this Service, it can not be used at the same time as `externalTrafficPolicy=Local`. Topology keys must be valid label keys and at most 16 keys may be specified. Endpoints are chosen based on the first topology key with available backends. If this field is specified and all entries have no backends that match the topology of the client, the service has no backends for that client and connections should fail. The special value "*" may be used to mean "any topology". This catch-all value, if used, only makes sense as the last value in the list. If `topologyKeys` is not specified or empty, no topology constraints will be applied.

**Comment by imroc** [comment]
Can I fire another PR on this? I can respond quickly to the review, help to make the service topology documentation gets merged before Docs deadline.

**Comment by johnbelamaric** [comment]
If you want that's fine with me, I am at KubeCon. Otherwise I can get to this later tonight or tomorrow morning.

**Comment by imroc** [comment]
Just using this PR then. I am a little busy too, try to take the time to write the doc.

**Comment by imroc** [comment]
## Introduction

Service Topology makes it possible for network traffic of service to be forwarded nearby, reducing network hops and improving network performance. It uses topology keys to decide the preferred topological domains to be used, e.g. node, rack, zone, region, etc.

**Comment by imroc** [comment]
## Using Topology Keys

``` yaml
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  type: ClusterIP
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  topologyKeys: ["kubernetes.io/hostname", "topology.kubernetes.io/zone", "*"]
```

please help to describe this 😅

**Comment by daminisatya** [comment]
@johnbelamaric Update on this PR?

**Comment by johnbelamaric** [comment]
I am working on this now and will have an update within the hour. Thanks for your patience, it's been a hectic week!

**Comment by johnbelamaric** [comment]
Ok, PTAL now. @imroc I also added the feature gate enabling for `kube-apiserver` - isn't that necessary for the API validation to kick in ?

**Comment by johnbelamaric** [comment]
/assign @andrewsykim @imroc 

Please review for technical content.

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+Nodes connected to the same top-of-rack switch for the lowest latency.
+
+## Prerequisites
+
+The following prerequisites are needed in order to enable topology aware service
+routing:
+
+   * Kubernetes 1.17 or later
+   * Kube-proxy running in iptables mode or IPVS mode
+   * Enable [Endpoint Slices](/docs/concepts/services-networking/endpoint-slices/) (enabled by default since v1.17)
```
Comment (robscott) [inline content/en/docs/concepts/services-networking/service-topology.md]: Sadly these are not enabled by default yet, although in a beta state. https://github.com/kubernetes/kubernetes/pull/85365

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+## Enable Service Topology
+
+To enable service topology, enable the `ServiceTopology` feature gate for
+kube-apiserver and kube-proxy:
```
Comment (robscott) [inline content/en/docs/concepts/services-networking/service-topology.md]: I think kube-controller-manager would also be key here for the sake of controller logic.

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+```
+--feature-gates="ServiceTopology=true"
+```
+
+## Using Service Topology
+
+If your cluster has Service Topology enabled, you can control Service traffic
+routing by specifying the `.spec.topologyKeys` field on the Service. This field
```
Comment (robscott) [inline content/en/docs/concepts/services-networking/service-topology.md]: nit: maybe "`topologyKeys` field on the Service spec".

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+Consider a cluster with Nodes that are labeled with their hostname, zone name,
+and region name, using the [well-known
+labels](docs/reference/kubernetes-api/labels-annotations-taints/). Then you can
+set the `topologyKeys` values of a service to direct traffic as follows.
+
+* Only to endpoints on the same node, failing if no endpoint exists on the node:
+  `["kubernetes.io/hostname"]`.
+* Preferentially to endpoints on the same node, falling back to endpoints in the
+  same zone, followed by the same region, and failing otherwise: `["kubernetes.io/hostname",
+  "failure-domain.beta.kubernetes.io/zone", "failure-domain.beta.kubernetes.io/region"]`.
```
Comment (robscott) [inline content/en/docs/concepts/services-networking/service-topology.md]: I believe this should be referencing the new topology labels used by EndpointSlices (`topology.kubernetes.io/zone` and `topology.kubernetes.io/region`): https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/endpointslice/utils.go#L60-L63.

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+  This may be useful, for example, in cases where data locality is critical.
+* Preferentially to the same zone, but fallback on any available endpoint if
+  none are available within this zone:
+  `["failure-domain.beta.kubernetes.io/zone", "*"]`.
```
Comment (robscott) [inline content/en/docs/concepts/services-networking/service-topology.md]: I think this should be `topology.kubernetes.io/zone` as well

File: content/en/docs/concepts/services-networking/service-topology.md (hunk 3 above)
Comment (johnbelamaric) [reply content/en/docs/concepts/services-networking/service-topology.md]: I don't think there is any controller logic for this. It's all done in kube-proxy.

Comment (robscott) [reply content/en/docs/concepts/services-networking/service-topology.md]: Woops, you're completely right, disregard that.

**Comment by johnbelamaric** [comment]
Thanks @robscott for the review! @daminisatya I think we are good to go now, assuming Rob can LGTM with the latest.

**Comment by robscott** [comment]
Thanks!

/lgtm

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+* The catch-all value, `"*"`, must be the last value in the topology keys, if
+  it is used.
+
+
+{{% /capture %}}
+
+{{% capture whatsnext %}}
+
+* Read about [enabling Service Topology](/docs/tasks/administer-cluster/enabling-service-topology)
```
Comment (imroc) [inline content/en/docs/concepts/services-networking/service-topology.md]: nit: enabling --> Enabling

And currently the doc Enabling Service Topology is not exists, will be added in later commit or another PR?

**Comment by johnbelamaric** [comment]
Oops...do we need it, since it’s in this  doc?

**Comment by imroc** [comment]
Maybe we don't need it, just add an simple Service example in this doc?

**Comment by daminisatya** [comment]
/lgtm 

@johnbelamaric Thank you for the PR. let me know once it is final, I will approve it.

**Comment by daminisatya** [comment]
@johnbelamaric Kindly update the PR with these small review comments ASAP!

File: content/en/docs/concepts/services-networking/service-topology.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
+---
+reviewers:
+- johnbelamaric
+- imroc
+title: Service Topology
+feature:
+  title: Service Topology
+  description: >
+    Routing of service traffice based upon cluster topology.
```
[the remark this answers is not in the document]
Comment (andrewsykim) [reply content/en/docs/concepts/services-networking/service-topology.md]: ^ I can fix this in a follow-up PR, I have some other changes I'd like to add :)

File: content/en/docs/concepts/services-networking/service-topology.md (hunk 7 above)
Comment (andrewsykim) [reply content/en/docs/concepts/services-networking/service-topology.md]: Added in this PR https://github.com/kubernetes/website/pull/17808 PTAL

**Comment by johnbelamaric** [comment]
Thanks all for following up on this, I was OOO and had very limited network access the last few days!

**Pull request merged**